common/http/src/retry.rs

```diff
@@ -44,88 +55,169 @@ struct DiskRead {
-                .and_then(|s| FromStr::from_str(s).ok())
-                .ok_or_else(|| Error::InvalidFileName(file_name.clone()))?;
-
-            if Utc::now().timestamp() - timestamp < self.retry_base_delay_secs {
-                continue;
+            if path.extension() == Some(std::ffi::OsStr::new("retry")) {
+                let file_name = path
+                    .file_name()
+                    .and_then(|s| s.to_str())
+                    .map(|s| s.to_string())
+                    .ok_or_else(|| Error::NonUtf8(path.clone()))?;
+
+                let timestamp: i64 = file_name
+                    .split('_')
+                    .map(|s| s.to_string())
+                    .collect::<Vec<String>>()
+                    .get(0)
+                    .and_then(|s| FromStr::from_str(s).ok())
+                    .ok_or_else(|| Error::InvalidFileName(file_name.clone()))?;
+
+                if Utc::now().timestamp() - timestamp < self.retry_base_delay_secs {
+                    continue;
+                }
+                self.waiting.push(path);
             }
-
-            self.waiting.push(path);
         }
 
         Ok(())
     }
 
-    fn read_from_disk(path: &Path) -> Result<(Option<Vec<Offset>>, IngestBody), Error> {
-        let mut file = File::open(path)?;
+    async fn read_from_disk(path: &Path) -> Result<(Option<Vec<Offset>>, IngestBody), Error> {
+        let mut file = BufReader::new(File::open(path).await?);
         let mut data = String::new();
-        file.read_to_string(&mut data)?;
-        remove_file(&path)?;
+        file.read_to_string(&mut data).await?;
+        remove_file(&path).await?;
         let DiskRead { offsets, body } = serde_json::from_str(&data)?;
         Ok((offsets, body))
     }
+
+    pub fn into_stream(
+        self,
+    ) -> impl Stream<Item = Result<(IngestBodyBuffer, Option<Vec<Offset>>), Error>> {
+        stream::unfold(self, |state| async move {
+            loop {
+                // Try to populate retry queue
+                if state.waiting.is_empty() {
+                    if let Err(e) = state.fill_waiting().await {
+                        return Some((Err(e), state));
+                    };
+                    // If there are still no objects then sleep for a while
+                    if state.waiting.is_empty() {
+                        Delay::new(Duration::from_secs(
+                            state.retry_base_delay_secs.try_into().unwrap_or(2),
+                        ))
+                        .await;
+                    }
+                }
+
+                if let Some(path) = state.waiting.pop() {
+                    match Retry::read_from_disk(&path).await {
+                        Ok((offsets, ingest_body)) => {
+                            match IntoIngestBodyBuffer::into(ingest_body).await {
+                                Ok(body_buffer) => {
+                                    return Some((Ok((body_buffer, offsets)), state))
```

Comment:
Should we throttle, e.g. in the order of 50-100ms, after a successful read to prioritize new logs?

Reply:
We could reintroduce the retry step delay config?

I accidentally pushed to the wrong branch so I've reopened the PR over at https://github.com/logdna/logdna-agent-v2/pull/222